Bugfix/76 flaky zerodivisionerror on linux tests

phargogh opened the pull request:
This PR addresses a flaky `ZeroDivisionError` that would occasionally be encountered on our ubuntu/psutil GHA tests (all python versions).  The root cause was that the graph (where `n_workers>=1`) could get into a state where:

1. A task is added
2. The task is started
3. The task fails, causing the graph to be terminated
4. Join is called sometime after graph termination, resulting in the exception passing without being propagated.

This silent passing of exceptions under these circumstances was inconsistent with how `TaskGraph.join()` was operating when there were incomplete tasks (some of which might raise an exception that _would_ propagate to the caller).  This PR changes this behavior to cause `TaskGraph.join()` to propagate the first exception encountered when the graph is joined.

Because of this change in behavior, it's possible for an exception to be raised twice: once when the erroring task is joined and the same exception again is raised when the graph is joined.  I think this is reasonable and more predictable behavior than silently passing an exception.

There's also some light linting, mostly of f-strings.

Fixes #76 

PS: Since this is a subtle but notable change in the behavior of TaskGraph, I tried running the InVEST test suite against this build just to make sure this wouldn't break everything.  Here are the test results (currently all passing): https://github.com/phargogh/invest/runs/3750631610?check_suite_focus=true

dcdenu4 commented on taskgraph/Task.py:
```diff
@@ -767,8 +766,12 @@ def join(self, timeout=None):
+        if self._n_workers < 0:
+            # Join() is meaningless since tasks execute synchronously.
+            LOGGER.debug(
+                'n_workers: %s; join is vacuously true' % self._n_workers)
```
Had to look up 'vacuous' truth!

phargogh replied:
haha I love this term

dcdenu4 commented on tests/test_task.py:
```diff
@@ -1238,21 +1242,24 @@ def test_return_value(self):
                         store_result=True,
                         args=(expected_value,),
-                        task_name=f'expected error {iteration_id}')
+                        task_name='expected error {iteration_id}')
```
```suggestion
                        task_name=f'expected error {iteration_id}')

```

dcdenu4 replied:
I think this still needs an `f`-string, no?

phargogh replied:
Whoops, that was sloppy of me!  Thanks for catching that.

dcdenu4 commented on tests/test_task.py:
```diff
@@ -1238,21 +1242,24 @@ def test_return_value(self):
 
-            task_graph.close()
-            task_graph.join()
+                with self.assertRaises(RuntimeError):
+                    task_graph.join()
```
We're expecting the `join` to now also propagate and raise the same exception from the above task? Can we avoid a `close` here because the `join` exception will handle that?

phargogh replied:
Oh interesting!  Do you mean in the general case or in this specific test?

In this test, I think you're right, closing the graph isn't strictly necessary.  The `task_graph.join()` will also terminate the graph when the exception is raised.  Because we only have one task to execute, we also happen to know that the whole graph has completed.

In general though, I think the `close` is still useful to have since it's possible to join the graph within the pipeline. 

```text

<define tasks>
graph.join()   <-- might raise exception

<define more tasks>
graph.close()   # No more tasks can be added after this
graph.join()  <-- also might raise exception
```

Do you think that makes sense here?  I'm a little wary of changing taskgraph behavior, so I want to be sure that this is a reasonable idea!

dcdenu4 replied:
Yea this makes sense, thanks. I was mostly curious about this specific case and the decision to drop the `task_graph.close()` call.

dcdenu4 commented on taskgraph/Task.py:
```diff
@@ -767,8 +766,12 @@ def join(self, timeout=None):
         """
         LOGGER.debug("joining taskgraph")
-        if self._n_workers < 0 or self._terminated:
+        if self._n_workers < 0:
```
Sorry, I'm a little confused trying to parse out how this working now. If `self._terminated` is `true`, the below `try/except` will raise the exception for the 2nd time? What part of that `try` do we expect to fail which will raise the exception?

phargogh replied:
> What part of that try do we expect to fail which will raise the exception?
It's the `task.join()` line.  Any exceptions raised within a task are then re-raised by `task.join()`.  So then when we do a `graph.join()`, any tasks with exceptions will have their errors propagated.

> If self._terminated is true, the below try/except will raise the exception for the 2nd time?
Well, it depends!  Let's say we have a task `E` that raises `ValueError`.

If we call `E.join()`, then that will raise `ValueError`.  When we call `graph.join()` at the end of the pipeline, `graph.join()` will _also_ raise `ValueError`.

But if we _don't_ call `E.join()` (which is what happens in most InVEST models), `E`'s exception will terminate the graph and log the exception, BUT the exception will never be raised.  This is the behavior that's changing in this PR ... if the graph is terminated, we still raise the error that caused it.

This second case is why we were getting the flaky `ZeroDivisionError` ... the task was starting up so fast that by the time we got to `graph.join()`, the error was logged but wouldn't be raised.  `graph.join()` was only raising the error if and only if there were any tasks remaining by the time `graph.join()` is called.

dcdenu4 replied:
>If we call E.join(), then that will raise ValueError. When we call graph.join() at the end of the pipeline, graph.join() will also raise ValueError.

>But if we don't call E.join() (which is what happens in most InVEST models), E's exception will terminate the graph and log the exception, BUT the exception will never be raised. This is the behavior that's changing in this PR ... if the graph is terminated, we still raise the error that caused it.

Brilliant, thanks for clarifying, that helped a lot.

dcdenu4 reviewed (changes requested):
Hey @phargogh, I mostly have a few clarifying questions and one suggestion to add back an f-string.

phargogh commented:
Your question in https://github.com/natcap/taskgraph/pull/77#discussion_r719335531 got me thinking and I'd like to test something else with this change before re-requesting your review.  More soon!

phargogh commented:
OK!  After testing this out a bit, I'm even more confident that this change in behavior is the right thing to do.

I modified Scenic Quality to raise an exception in an intermediate task and also to simulate a long time before join is called I added `time.sleep(30)` just before the end of `execute`.

```diff
diff --git a/src/natcap/invest/scenic_quality/scenic_quality.py b/src/natcap/invest/scenic_quality/scenic_quality.py
index 41d5e1a8..264557be 100644
--- a/src/natcap/invest/scenic_quality/scenic_quality.py
+++ b/src/natcap/invest/scenic_quality/scenic_quality.py
@@ -404,6 +404,8 @@ def execute(args):
     )

     LOGGER.info('Waiting for Scenic Quality tasks to complete.')
+    import time
+    time.sleep(30)
     graph.join()


@@ -898,6 +900,8 @@ def _count_and_weight_visible_structures(visibility_raster_path_list, weights,
     dem_raster_info = pygeoprocessing.get_raster_info(clipped_dem_path)
     dem_nodata = dem_raster_info['nodata'][0]

+    raise Exception('Foo!')
+
     pygeoprocessing.new_raster_from_base(
         clipped_dem_path, target_path, gdal.GDT_Float32, [target_nodata],
         raster_driver_creation_tuple=FLOAT_GTIFF_CREATION_OPTIONS)
``` 

With `taskgraph==0.10.3`, this change will actually result in "Model Success"!  The reason for this is that the exception from an already-completed task is not propagated once `graph.join()` is reached.

By contrast, installing the latest rev on this branch, the `Exception('Foo!')` _is_ propagated on `graph.join()` and the exception is raised.

In InVEST, the former situation was unlikely to be triggered because, in general, our models probably tend to arrive at `graph.join()` fairly quickly, minimizing the chance that we'd end up in this intermediate state.

So, as a result, I'm much more confident that this behavioral change is the right one.

dcdenu4 commented:
> The reason for this is that the exception from an already-completed task is not propagated once graph.join() is reached.

> By contrast, installing the latest rev on this branch, the Exception('Foo!') is propagated on graph.join() and the exception is raised.

Woah! Very interesting, thanks for diving into that and explaining here. You've answered all my questions so I think this is good to go!

dcdenu4 reviewed (approved):
Nice!
